root/dataframe/test_trainCache.C

```diff
@@ -98,7 +98,57 @@ int WithTTreeReader()
    chain.GetTree()->PrintCacheStats();
 
    if (refNBytes != rbRAII.GetBytesRead()) {
-      std::cout << "With TChain: the bytes read (" << rbRAII.GetBytesRead()
+      std::cout << "With WithTTreeReader: the bytes read (" << rbRAII.GetBytesRead()
+                << ") are different from the reference (" << refNBytes << ")\n";
+      ret +=1;
+   }
+
+   return ret;
+}
+
+int WithTTreeReaderAndTwoBranches()
+{
+   std::cout << "\n----------- With TTreeReaderAndTwoBranches" << std::endl;
+
+   TChain chain(treeName);
+   chain.Add(fileName);
+
+   TTreeReader reader(&chain);
+
+   // We read branch 2 to try to confuse the cache
+   reader.SetEntriesRange(firstEvt, lastEvt);
+   TTreeReaderValue<double> vd2(reader, branchName2);
+
+   while(reader.Next()){
+      *vd2;
+   }
+
+   // Now we restart, attach a new reader and read the branch
+   reader.Restart();
```

Comment:
In addition to restarting, I would also test with a distinct TTreeReader object.